Feature/detectn

iwillspeak opened the pull request:
Proposal for adding length based string API for detect.


Fixes #7.

iwillspeak committed:
Bump Whatlang Package Dependency

iwillspeak committed:
Add `whatlang_detectn` Overload

This overload allows a user to pass in a length explicitly rather than relying on null termination of a string.

iwillspeak committed:
Return Length of Language Names

Update the API to allow callers to know the length of the string
copied to an output buffer.

Dr-Emann commented on src/lib.rs:
```diff
@@ -40,38 +45,66 @@ pub extern fn whatlang_detect(ptr: *const c_char, cinfo: &mut CInfo) -> u8 {
     }
 }
 
+fn detect_internal(text: &str, cinfo: &mut CInfo) -> u8 {
```
Since both users of `detect_internal` actually have &[u8] available, and do the same handling for utf8 parsing, maybe this function should take &[u8], and handle parsing into a str and returning `2` itself?

iwillspeak replied:
Good shout. I’ll look at making that edit.

Dr-Emann commented on src/lib.rs:
```diff
@@ -40,38 +45,66 @@ pub extern fn whatlang_detect(ptr: *const c_char, cinfo: &mut CInfo) -> u8 {
     unsafe {
-        libc::strcpy(buffer_ptr, s.as_ptr());
+		if buffer_ptr != ptr::null_mut() {
+			libc::strncpy(buffer_ptr, s.as_ptr(), 30);
```
`strncpy` is almost never what you want, since it zero-fills up to `n` bytes, and doesn't always null terminate (if `s` is 30 bytes long).

How about something like:

```rust
unsafe fn copy_to_cstr(src: &str, dst: *mut c_char) {
    let len = src.len();
    let src = src.as_ptr().cast::<c_char>();
    src.copy_to_nonoverlapping(dst, len);
    *dst.add(len) = 0;
}
```

Then all of these functions (`eng_name`, `code`, `name`, etc) can be something like

```rust
pub extern "C" fn whatlang_lang_eng_name(lang: Lang, buffer_ptr: *mut c_char) -> size_t {
    let s = lang.eng_name();
    if !buffer_ptr.is_null() {
        unsafe { copy_to_cstr(s, buffer_ptr) };
    }
    s.len()
}
```

This avoids an extra allocation with `CString`.

iwillspeak replied:
Nice. Thinking we could maybe move the buffer null check into the copy too.

Dr-Emann commented on src/whatlang.h:
```diff
@@ -120,11 +120,12 @@ struct whatlang_info {
-void whatlang_lang_eng_name(uint8_t lang, char* buffer);
-void whatlang_lang_name(uint8_t lang, char* buffer);
-void whatlang_script_name(uint8_t script, char* buffer);
+size_t whatlang_lang_code(uint8_t lang, char* buffer);
```
Should probably add some documentation about what the return value means.

iwillspeak replied:
:+1:

iwillspeak committed:
Address Code Review Comments

Move more shared work into `detect_internal`. Simplify `copy_cstr` so
we no longer need to allocate. Add documentation for the C API.

Fixes #3.

iwillspeak commented:
Ok. I think I have addressed those comments.

Dr-Emann commented on src/lib.rs:
```diff
@@ -14,64 +16,71 @@ pub struct CInfo {
     }
 }
+
+fn copy_cstr(src: &str, dst: *mut c_char) -> size_t {
```
Can we make this function unsafe, as well as all the extern "C" fns?

iwillspeak replied:
Unsafe on all the extern functions. Got rid of the unneeded unsafe blocks.

Dr-Emann commented on src/lib.rs:
```diff
@@ -14,64 +16,71 @@ pub struct CInfo {
+pub extern "C" fn whatlang_detectn(ptr: *const c_char, len: libc::size_t, cinfo: &mut CInfo) -> u8 {
+    let text = unsafe {
+        core::slice::from_raw_parts(ptr as *const u8, len)
+    };
+	detect_internal(&text, cinfo)
```
There's a couple of places where the spacing looks off here. I'm not sure of this repo's policy for formatting, it looks like `cargo fmt` will touch a few other places than what you've changed, but I'd personally prefer sticking to cargo fmt's choices.

iwillspeak replied:
Reformatted the whole thing.

iwillspeak committed:
Address More Code Review Comments

Reformat with rustfmt. Move extern FNs to be `unsafe` and mark
the `copy_cstr` as unsafe overall.

iwillspeak committed:
Expose Status Codes to C API

Update C API to expose the return status as a C style Enum. Check the
input pointer for NULL too.

Fixes: #5 #1

iwillspeak commented on src/whatlang.h:
```diff
@@ -5,6 +5,13 @@
 #include <stdlib.h>
 
+typedef enum whatlang_status : uint8_t {
+    OK = 0,
```
These enum values need prefixing. We should also switch the constants over to being enums too.

iwillspeak replied:
Improved enum representation now. Users can choose between const definitions or clang / C++11 compatible uint8_t enums.

iwillspeak committed:
Allow the User to Choose Enum Type

Expose the choice of constnts vs enums to the user with the
`WHATLANG_USE_C_ENUMS` compiler definition. When this is defined
instead of constants the header file will define enums. We typedef for
backwards compatibility so all use sites should be backwards
compatible.

Fixes: #4.

iwillspeak committed:
Add Simple Useage Docs to README

Include a quick overview of the use of this library, along with
discussion of the `WHATLANG_USE_C_ENUMS` feature.

greyblake commented:
@iwillspeak @Dr-Emann I'm quite surprised, because I thought nobody uses whatlang-ffi. I am curious where do you use it?
Thanks for you contribution! And thanks @Dr-Emann for reviewing it!

iwillspeak commented:
I’m currently evaluating whatlang for use from .net. Hopefully the dotnet bindings I have written can be open sourced at some point.

greyblake commented:
@iwillspeak wow, that is interesting. Hope that it will work for you out!

iwillspeak commented:
I hope it will work out too. I don’t suppose I could be cheeky and ask you tag this PR as `hacktoberfest-accepted`?

greyblake commented:
@iwillspeak I added `hacktoberfest-accepted` label

Dr-Emann commented on src/whatlang.h:
```diff
@@ -5,126 +5,187 @@
-static const uint8_t WHATLANG_LANG_YDD = 80;
-static const uint8_t WHATLANG_LANG_YOR = 81;
-static const uint8_t WHATLANG_LANG_ZUL = 82;
+#ifdef WHATLANG_USE_C_ENUMS
+# define _C_ENUM_THUNK(key, val) key = val,
+# define DEFINE_ENUM(name, cb)			\
+	typedef enum name : uint8_t {	    \
```
Despite the name `WHATLANG_USE_C_ENUMS`, this creates C++ enums: C (unfortunately) still doesn't allow specifying the underlying type of an enum.

Dr-Emann replied:
@iwillspeak @greyblake

iwillspeak commented:
It worked in plain old clang, but not gcc. It’s a clang extension I assume.

On Mon, 18 Oct 2021 at 17:47, Zachary Dremann ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/whatlang.h
> <https://github.com/greyblake/whatlang-ffi/pull/8#discussion_r731124613>:
>
> > +#ifdef WHATLANG_USE_C_ENUMS
> +# define _C_ENUM_THUNK(key, val) key = val,
> +# define DEFINE_ENUM(name, cb)			\
> +	typedef enum name : uint8_t {	    \
>
> @iwillspeak <https://github.com/iwillspeak> @greyblake
> <https://github.com/greyblake>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greyblake/whatlang-ffi/pull/8#discussion_r731124613>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHVG4KYLNWQ6TEPQVKKIKDUHRFQTANCNFSM5FN5CUHQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

Dr-Emann commented:
Looks that way: https://c.godbolt.org/z/4fEPKn4TM
